File: GFG/Print leaf nodes from preorder traversal of BST/Print leaf nodes from preorder traversal of BST.java
class Solution
{
    public int[] leafNodes(int arr[], int N)
    {
        // code here
        ArrayList<Integer> ans=new ArrayList<>();
        Stack<Integer> st=new Stack<>();
        st.push(arr[0]);
        for(int i=1;i<N;i++){
            if(arr[i]<st.peek()){
                st.push(arr[i]);
            }else{
                int temp=st.peek();
                int removed=0;
                while(!st.isEmpty() && st.peek()<arr[i]){
                    st.pop();
                    removed++;
                }
                st.push(arr[i]);
                if(removed>=2){
                    ans.add(temp);
                }
            }
        }
        ans.add(st.peek());
        int[] res=new int[ans.size()];
        int idx=0;
        for(int i:ans){
            res[idx++]=i;
        }
        return res;
    }
}

File: GFG/Print leaf nodes from preorder traversal of BST/Print leaf nodes from preorder traversal of BST.cpp
class Solution {
  public:
    vector<int> leafNodes(int a[],int n) {
        // code here
        
         stack<int> s;
        vector<int> vec;
        s.push(a[0]);
        for(int i=1;i<n;i++){
            if(s.top()<a[i]){
                int pc = 0;
                int ln = s.top();
                while(!s.empty()&&s.top()<a[i]){
                    s.pop();
                    pc++;
                }
                if(pc>1){
                    vec.push_back(ln);
                }
            }
            s.push(a[i]);
        }
        if(!s.empty()){
            vec.push_back(s.top());
        }
        return vec;
    }
};
